### PR TITLE
OCPBUGS-20085: IBMCloud: Handle disk delete errors

### PR DESCRIPTION
**Pull request by cjschaef**
Handle cases when IBM Cloud disk deletion returns an error, without any response data.

### CODE REVIEW

**Comment by cjschaef** [comment]
@sadasu 
Yeah, I will have to get a Jira opened to get this backported. I'll see if I can compile the details and do that, and retitle this with that OCPBUG once I have it.

cc @MayXuQQ I'll take your data to open the bug, but reproducing the error will likely be hard, since it is a very unique case from IBM Cloud API's.

**Comment by MayXuQQ** [comment]
@cjschaef I just found this bug once

**Comment by MayXuQQ** [comment]
found this issue in https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.13-amd64-nightly-ibmcloud-ipi-rt-f28/1710327579702988800
version: 4.13.0-0.nightly-2023-10-05-083326 

> level=info msg=Deleted instance "ci-op-lg1g50r7-6990f-fhmz9-worker-3-b9645"
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=3, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=4, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=5, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=6, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=7, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=8, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
level=info msg=UNEXPECTED RESULT, Re-attempting execution .., attempt=9, retry-gap=10, max-retry-Attempts=30, stopRetry=false, error=<nil>
E1006 23:21:22.092062      37 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 188 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x6309820?, 0x229b65d0})
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0001182a0?})
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75
panic({0x6309820, 0x229b65d0})
	/usr/lib/golang/src/runtime/panic.go:884 +0x212
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).waitForDiskDeletion.func1()
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:84 +0x12a
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).Retry(0xc0001c8b00, 0xc0005316e8)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:99 +0x73
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).waitForDiskDeletion(0xc0001c8b00, {{0xc00159a2a0, 0x29}, {0xc00159a2d0, 0x28}, {0xc000c7a624, 0x9}, {0x7c72f73, 0x4}, {0xc00159a2a0, ...}})
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:78 +0x14f
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).destroyDisks(0xc0001c8b00)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:118 +0x508
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).executeStageFunction.func1()
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:201 +0x3f
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x18, 0xc000580000})
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:222 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x1dde5418?, 0xc000130000?}, 0xc000487e90?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:235 +0x57
k8s.io/apimachinery/pkg/util/wait.poll({0x1dde5418, 0xc000130000}, 0xd0?, 0x132ea05?, 0x30?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:582 +0x38
k8s.io/apimachinery/pkg/util/wait.PollImmediateInfiniteWithContext({0x1dde5418, 0xc000130000}, 0x4130a7?, 0x28?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:568 +0x49
k8s.io/apimachinery/pkg/util/wait.PollImmediateInfinite(0xc000916718?, 0xc000487f70?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:557 +0x46
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).executeStageFunction(0xc0001c8b00, {{0x7c73c5c?, 0xc000487fd0?}, 0xc001672140?}, 0x53eaaa?, 0xc00088001c?)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:198 +0x108
created by github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).destroyCluster
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:172 +0xba5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x537bdea]

goroutine 188 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc0001182a0?})
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7
panic({0x6309820, 0x229b65d0})
	/usr/lib/golang/src/runtime/panic.go:884 +0x212
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).waitForDiskDeletion.func1()
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:84 +0x12a
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).Retry(0xc0001c8b00, 0xc0005316e8)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:99 +0x73
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).waitForDiskDeletion(0xc0001c8b00, {{0xc00159a2a0, 0x29}, {0xc00159a2d0, 0x28}, {0xc000c7a624, 0x9}, {0x7c72f73, 0x4}, {0xc00159a2a0, ...}})
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:78 +0x14f
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).destroyDisks(0xc0001c8b00)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/disk.go:118 +0x508
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).executeStageFunction.func1()
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:201 +0x3f
k8s.io/apimachinery/pkg/util/wait.ConditionFunc.WithContext.func1({0x18, 0xc000580000})
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:222 +0x1b
k8s.io/apimachinery/pkg/util/wait.runConditionWithCrashProtectionWithContext({0x1dde5418?, 0xc000130000?}, 0xc000487e90?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:235 +0x57
k8s.io/apimachinery/pkg/util/wait.poll({0x1dde5418, 0xc000130000}, 0xd0?, 0x132ea05?, 0x30?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:582 +0x38
k8s.io/apimachinery/pkg/util/wait.PollImmediateInfiniteWithContext({0x1dde5418, 0xc000130000}, 0x4130a7?, 0x28?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:568 +0x49
k8s.io/apimachinery/pkg/util/wait.PollImmediateInfinite(0xc000916718?, 0xc000487f70?)
	/go/src/github.com/openshift/installer/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:557 +0x46
github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).executeStageFunction(0xc0001c8b00, {{0x7c73c5c?, 0xc000487fd0?}, 0xc001672140?}, 0x53eaaa?, 0xc00088001c?)
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:198 +0x108
created by github.com/openshift/installer/pkg/destroy/ibmcloud.(*ClusterUninstaller).destroyCluster
	/go/src/github.com/openshift/installer/pkg/destroy/ibmcloud/ibmcloud.go:172 +0xba5

**Comment by sadasu** [comment]
@cjschaef what do the errors in e2e-ibmcloud-ovn indicate?

**Comment by sadasu** [comment]
/approve

e2e-ibmcloud-ovn has been perma-failing and this fix has not changed that.

**Comment by cjschaef** [comment]
@sadasu Sorry for the delay, yes the major failure I have seen is the event check for `csi-drivers`, which appears to occur frequently on 4.15.

I haven't gotten a chance to open a Jira, but have notified IBM Cloud Storage to investigate further, likely cleaning up the repeated events causing the failure.

I am hoping the other failures are flakes, but I'll try another few times to see if everything clears up and get an issue open for `events should not repeat pathologically for ns/openshift-cluster-csi-drivers`

**Comment by cjschaef** [comment]
Known CSI driver bug for IBM Cloud opened
https://issues.redhat.com/browse/OCPBUGS-22331

/test e2e-ibmcloud-ovn

**Review by sadasu — APPROVED** [review]

File: pkg/destroy/ibmcloud/disk.go [hunk 1]
```diff
@@ -58,14 +58,16 @@ func (o *ClusterUninstaller) deleteDisk(item cloudResource) error {
 	options := o.vpcSvc.NewDeleteVolumeOptions(item.id)
 	details, err := o.vpcSvc.DeleteVolumeWithContext(ctx, options)
 
-	if err != nil && details.StatusCode != http.StatusNotFound {
-		return errors.Wrapf(err, "Failed to delete disk name=%s, id=%s.If this error continues to persist for more than 20 minutes then please try to manually cleanup the volume using - ibmcloud is vold %s", item.name, item.id, item.id)
-	}
+	if err != nil {
+		if details == nil || details.StatusCode != http.StatusNotFound {
+			return errors.Wrapf(err, "Failed to delete disk name=%s, id=%s.If this error continues to persist for more than 20 minutes then please try to manually cleanup the volume using - ibmcloud is vold %s", item.name, item.id, item.id)
```
Comment (barbacbd) [inline pkg/destroy/ibmcloud/disk.go:63]:
```suggestion
			return fmt.Errorf("Failed to delete disk name=%s, id=%s.If this error continues to persist for more than 20 minutes then please try to manually cleanup the volume using - ibmcloud is vold %s: %w", item.name, item.id, item.id, err)
```

Comment (cjschaef) [reply pkg/destroy/ibmcloud/disk.go]: Fixed

**Review by barbacbd — COMMENTED** [review]
/lgtm

**Pull request merged**